File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_cost = prices[-1]
        result = 0
        for i in range(len(prices) - 1, -1, -1):
            if prices[i] < max_cost:
                result = max(result, max_cost - prices[i])
            if prices[i] > max_cost:
                max_cost = prices[i]
        return result

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int max = prices[n - 1];
        int result = 0;

        for (int i = n - 1; i > -1; i--) {
            if (prices[i] < max) {
                result = Math.max(result, max - prices[i]);
            } else {
                max = prices[i];
            }
        }

        return result;
    }
}